tests/Authenticator.hs

```diff
@@ -0,0 +1,426 @@
+  M.ClientDataHash ->
+  -- | rpEntity: The Relying Party's PublicKeyCredentialRpEntity.
+  M.PublicKeyCredentialRpEntity ->
+  -- | userEntity: The user account’s PublicKeyCredentialUserEntity, containing
+  -- the user handle given by the Relying Party.
+  M.PublicKeyCredentialUserEntity ->
+  -- | requireResidentKey: The effective resident key requirement for
+  -- credential creation, a Boolean value determined by the client.
+  -- NOTE: We always provide resident keys
+  Bool ->
+  -- | requireUserPresence: The constant Boolean value true. It is included
+  -- here as a pseudo-parameter to simplify applying this abstract
+  -- authenticator model to implementations that may wish to make a test of
+  -- user presence optional although WebAuthn does not.
+  -- NOTE: We currently always have user present
+  Bool ->
+  -- | requireUserVerification: The effective user verification requirement for
+  -- credential creation, a Boolean value determined by the client.
+  -- NOTE: We currently always verify the user
+  Bool ->
+  -- | credTypesAndPubKeyAlgs: A sequence of pairs of PublicKeyCredentialType
+  -- and public key algorithms (COSEAlgorithmIdentifier) requested by the
+  -- Relying Party. This sequence is ordered from most preferred to least
+  -- preferred. The authenticator makes a best-effort to create the most
+  -- preferred credential that it can.
+  [M.PublicKeyCredentialParameters] ->
+  -- | excludeCredentialDescriptorList: An OPTIONAL list of
+  -- PublicKeyCredentialDescriptor objects provided by the Relying Party with
+  -- the intention that, if any of these are known to the authenticator, it
+  -- SHOULD NOT create a new credential. excludeCredentialDescriptorList
+  -- contains a list of known credentials.
+  [M.PublicKeyCredentialDescriptor] ->
+  -- | enterpriseAttestationPossible: A Boolean value that indicates that
+  -- individually-identifying attestation MAY be returned by the authenticator.
+  Bool ->
+  -- | extensions: A CBOR map from extension identifiers to their authenticator
+  -- extension inputs, created by the client based on the extensions requested
+  -- by the Relying Party, if any.
+  Maybe M.AuthenticationExtensionsClientInputs ->
+  m (M.AttestationObject, Authenticator)
+authenticatorMakeCredential
+  authenticator@AuthenticatorNone {..}
+  _hash
+  rpEntity
+  userEntity
+  _requireResidentKey
+  _requireUserPresence
+  _requireUserVerification
+  credTypesAndPubKeyAlgs
+  excludeCredentialDescriptorList
+  _enterpriseAttestationPossible
+  _extensions =
+    do
+      -- 1. Check if all the supplied parameters are syntactically well-formed
+      -- and of the correct length. If not, return an error code equivalent to
+      -- "UnknownError" and terminate the operation.
+      -- NOTE: This step is performed during decoding, the only thing we can do
+      -- is to make sure we have a RpId
+      let rpId = fromMaybe (M.RpId "localhost") $ M.pkcreId rpEntity
```

Comment:
From this and also from the client implementation it looks like the origin/domain/relying party we are connected to should be a parameter, such that we don't need to hardcode `localhost`. Note that the default relying party can be derived from the domain, see the note in https://www.w3.org/TR/webauthn-2/#rp-id

Reply:
It is the client that has to perform this translation, so I propose allowing the authenticator to fail on a Nothing RpId, and adding code to the client that ensures it is not Nothing. This also keeps the authenticator conforming to spec, which is nice.

Reply:
That sounds good